### DISCUSSION TITLE
Proposal: contentLink Enhancement (Revised)

### DISCUSSION THREAD

**Discussion by nathangathright** [Podcastindex-org/podcast-namespace#614] [Enhancement Proposal]
### Purpose
`contentLink` was [initially proposed](https://github.com/Podcastindex-org/podcast-namespace/issues/325#issuecomment-1013467533) in the context of `liveItem` to allow podcasters to provide external destinations where the same livestream can be consumed. Expanding that to include support for on-demand episodes would enable podcasters to link to alternate versions of the same episode, such as a video version on YouTube or an extended, ad-free version on Patreon. Apps that support this tag could treat it like an external version of `alternateEnclosure`.

### Parent
`<item>` or `<podcast:liveItem>`

### Count
Multiple

### Node Value
The node value is a free-form string meant to explain to the user where this content link points and/or the nature of its purpose.

### Attributes
* **href:** (required) A URL pointing to an external platform representing the same content as the parent tag.

#### Item-level contentLink
```xml
<podcast:contentLink href="https://patreon.com/posts/public-episode-83079838">Listen on Patreon</podcast:contentLink>
<podcast:contentLink href="https://youtube.com/watch?v=RWPmCpKVPS8&list=PLKLAQeES5doHtEgyGyAInyDDlG-MljkMX">Watch on YouTube</podcast:contentLink>
```

**Comment by daveajones** [comment]
This seems like a simple change that would add a lot of flexibility.

**Reply by johnspurlock** [reply]
agreed, only suggestion would be to add a clear item-level example to the spec for folks wanting to associate a podcast episode with the associated youtube video watch page - for easy copy/paste

**Reply by nathangathright** [reply]
The current example YouTube URL is pointing to the watch page for a specific video: `https://www.youtube.com/watch?v=RWPmCpKVPS8` + the podcast playlist the video belongs too `&list=PLKLAQeES5doHtEgyGyAInyDDlG-MljkMX`. Should we remove the playlist bit to clarify that item-level contentLinks shouldn’t point to "playlists"?

**Reply by johnspurlock** [reply]
yea when the time comes to draft a PR for this (maybe it's worth doing now?), I'm suggesting keeping the new item-level youtube example in the spec as simple to follow as possible, leave little to the wild imagination...

**Reply by johnspurlock** [reply]
Dave added the new parent element to the NS, here is a quick followup PR with that item-level example and other corresponding fixes: https://github.com/Podcastindex-org/podcast-namespace/pull/652

**Comment by geeknews** [comment]
If this can be approved and put in a release stack this would help fix the pending clash of the Alternate Enclousre

**Reply by samsethi** [reply]
This doesn't work for me.  These are outbound links.  So for Geek News you could use the ContentLink to point to your YouTube channel.  As an app developer I want you to bring your video to my app not send you from my app to YouTube so we won't support it in TrueFans. If we add the Youtube Share link to the Alt Enclosure we can show the mime/type MP4 and also the label i.e YT video.  The contentLink has no attributes like that for us.

**Reply by nathangathright** [reply]
@samsethi If we really need attributes, then I’d agree that `alternateEnclosure` would be the better home for these links. Maybe I’m not deep enough in the weeds, and I should defer to your experience here, but as I understand things, displaying YouTube embeds only really requires the YouTube URL. Am I mistaken?

- If the podcaster adds a YouTube link to their feed, it’s not a direct link to a media file, like all other enclosure sources. So why would a video MIME type be appropriate? Apps are expecting an MP4 video from sources with that MIME type. If you’re going to iframe in the content, wouldn’t a MIME type of `text/html` be more appropriate?
- Since it’s not a direct link to a media file, I assume sources like these require a special case based on the domain name of the source to generate an appropriate iframe embed/HLS player. Why does the podcaster need to specify a MIME type at all?
- After navigating all that, I don’t understand how `alternateEnclosure` is a better home for non-media file source links.

And to reiterate what I said on Mastodon: No one is asking you to send people away from your app. Supporting a tag called "contentLink" doesn’t require displaying it as an external link. It might just be a more semantically appropriate place for feeds to provide these URLs, but no one can dictate the experience in your app besides you.

**Reply by jamescridland** [reply]
An alternateEnclosure is a link to a media file that plays the same (or similar) audio/video. It is absolutely suitable for TrueFans to ingest.

A contentLink is a tool for directories and search engines to help link a visitor to the platform of their choice. It is absolutely not suitable for TrueFans to use - why would you link away from your platform?

I don't see the clash here, but also see this as being (the first?) tag that TrueFans won't support, and I'd understand that entirely. Podnews would consume it.

**Reply by nathangathright** [reply]
> An alternateEnclosure is a link to a media file that plays the same (or similar) audio/video. It is absolutely suitable for TrueFans to ingest.

@jamescridland I 100% agree with this, which is why I’m confused that we’re claiming that a YouTube link is a media file. Adding them to an `alternateEnclosure` with `type=mp4` isn’t appropriate for these links. Please help me understand.

> A contentLink is a tool for directories and search engines to help link a visitor to the platform of their choice. It is absolutely not suitable for TrueFans to use - why would you link away from your platform?

I’ve been very clear that I don’t expect players to link out to competing players.

If players want to embed YouTube content in their app and we agree that a YouTube link is not a media file, then I’m open to any suggestions for the best way for feeds to provide these YouTube URLs that doesn’t break existing implementations of alternateEnclosure.

**Reply by jamescridland** [reply]
A YouTube link isn't a media file. It's also incorrect to add a mimeType of "video/mp4" to it, because you're not linking to a media file. It's not an alternateEnclosure, to my mind.

But it could be a special case - if you see "youtube.com" in an alternateEnclosure (with a `v` attribute), then you know how to deal with it. I'd suggest it simply doesn't have a mimeType because it isn't a file. (Or, text/html if you must). Right now, the mimeType ought to be parsed by a player and it should know which mimeTypes it can play and which it can't. I'd hope no iOS device is using my `audio/opus` alternateEnclosure, as one example, because it won't play it.

So it might be simply that: "If an alternateEnclosure does not have a mimeType selected, then the player should make a decision on how to deal with that URL by itself".

There is, of course, the `<link>` tag, which *could*, I guess, have a `rel="youtube"` attribute?

The media namespace allows links to embedded players. It's a namespace used by YouTube for different-sized episode images. It's supported by Omny Studio and others. But, this is an audio player and not a video one, so... I don't really know what to suggest.

```
<media:content url="https://traffic.omny.fm/d/clips/ad4ac715-121e-4de7-98e9-a5da0072a7a0/8d94b967-a751-481f-a663-a5da0082be6b/18de2768-6648-4b6e-acd7-b14e014806e4/audio.mp3?utm_source=Podcast&in_playlist=07a7a7f4-0393-4915-8753-a5da0086dece" type="audio/mpeg">
<media:player url="https://omny.fm/shows/game-changers-radio/the-more-the-ferrier/embed"/>
</media:content>
```

**Reply by nathangathright** [reply]
@samsethi For the last time, no one is asking apps to link out to competitors.

I hope TrueFans moves away from requiring feeds to use `type="video/mp4"` to enable this feature cause I think it’d be a shame to break existing implementations of alternateEnclosure. Either adopt something like `type="text/html"`, propose a new attribute, or ignore attributes altogether and parse the URL for YouTube URLs.

**Reply by jamescridland** [reply]
> Instead we want users to click an external content link

No, we don't. But we do want to help app developers understand that this YouTube link is not a link to a `video/mp4`.

**Reply by Marzal** [reply]
Related discussion and poll on the PSP
https://github.com/Podcast-Standards-Project/proposals/issues/2

**Comment by daveajones** [comment]
This change has been formalized.
